063-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,208 @@
+
+Note that directly instantiating metric reporters in the plugin is a workaround as Kafka currently does not provide a way for plugins to register metrics. [KIP-877](https://cwiki.apache.org/confluence/display/KAFKA/KIP-877%3A+Mechanism+for+plugins+and+connectors+to+register+metrics), currently in discussion, aims at solving this issue. Once this KIP is accepted, we should update the plugin to use this mechanism.
+
+### strimzi-kafka-bridge
+
+To function, the bridge instantiates Kafka clients. Today it has its own custom mechanism to retrieve metrics from the Kafka clients via JMX and expose them to Prometheus.
+
+We can update the bridge to use metric reporters to retrieve metrics from Kafka clients and still keep it's own mechanism to expose them to Prometheus (as it's also exposing its own metrics). To switch behavior, I propose introducing a new configuration to the bridge `metricsMode`/`KAFKA_BRIDGE_METRICS_MODE` which would initially default to `jmx` to keep the current behavior. If set to `reporter`, the bridge would set the `metric.reporters` configuration on all Kafka clients it starts to `KafkaPrometheusMetricsReporter` and retrieve their metrics via the Prometheus metrics registry. It should also set `prometheus.metrics.reporter.listener` to `disabled` so `KafkaPrometheusMetricsReporter` instances don't start their own HTTP endpoint.
+
+### kafka-quotas-plugin
+
+This plugin creates metrics by using the Yammer metrics library. `com.yammer.metrics.Metrics` registers metrics in the default Yammer registry and automatically exports metrics via JMX.
+
+Like strimzi-kafka-oauth, this plugin should be updated once [KIP-877](https://cwiki.apache.org/confluence/display/KAFKA/KIP-877%3A+Mechanism+for+plugins+and+connectors+to+register+metrics) is available in Kafka. 
+
+In the meantime, `YammerPrometheusMetricsReporter` will retrieve metrics from both the Kafka Yammer registry and the default Yammer registry to get all Yammer metrics.
+
+## Compatibility
+
+Differences with jmx_exporter metrics:
+
+- The reporter does not compute 1/5/15 minute rate, mean, max, min, stddev metrics. It's preferable to compute them in Prometheus instead of on the client side.
+- The reporter is missing the `kafka_server_app_info_starttimems` metric with the client/broker id label. (Due to [KAFKA-15186](https://issues.apache.org/jira/browse/KAFKA-15186))
+- Kafka exposes some non-numeric metrics. Prometheus only supports numeric values for metrics. Using jmx_exporter it's possible with rules to move the values into labels and still retrieve them in Prometheus. With this proposal non-numeric metrics will be ignored and not exposed. I plan to raise a KIP in Kafka to provide alternative to non-numeric metrics. For example there is already [KIP-972](https://cwiki.apache.org/confluence/display/KAFKA/KIP-972%3A+Add+the+metric+of+the+current+running+version+of+kafka) in progress to address some of them.
```

Comment:
I thought we were moving non-numerics to labels (as in Handling of non-numeric metrics section)?

Reply:
Yes this paragraph should be removed now. Done